### DISCUSSION TITLE
Wasm App Do Not Start On Nightly Build

### DISCUSSION THREAD

**Discussion by cangunaydin** [AvaloniaUI/Avalonia#14821] [Q&A]
## Describe the bug
Hello when i want to try the latest nightly build and test it on web assembly, on console it throws an error

MONO_WASM: Failed to load config file ./blazor.boot.json SyntaxError: Unexpected end of JSON input SyntaxError: Unexpected end of JSON input at https://localhost:5001/dotnet.js:3:25380 at https://localhost:5001/dotnet.js:3:25808 at async https://localhost:5001/dotnet.js:3:25082 at async https://localhost:5001/dotnet.js:3:24942 at async Object.create (https://localhost:5001/dotnet.js:3:34517) at async https://localhost:5001/main.js:6:23 Error: Failed to load config file ./blazor.boot.json SyntaxError: Unexpected end of JSON input SyntaxError: Unexpected end of JSON input at https://localhost:5001/dotnet.js:3:25380 at https://localhost:5001/dotnet.js:3:25808 at async https://localhost:5001/dotnet.js:3:25082 at async https://localhost:5001/dotnet.js:3:24942 at async Object.create (https://localhost:5001/dotnet.js:3:34517) at async https://localhost:5001/main.js:6:23 at https://localhost:5001/dotnet.js:3:26612 at async https://localhost:5001/dotnet.js:3:24942 at async Object.create (https://localhost:5001/dotnet.js:3:34517) at async https://localhost:5001/main.js:6:23
dotnet.js:3Uncaught Error: Failed to load config file ./blazor.boot.json SyntaxError: Unexpected end of JSON input SyntaxError: Unexpected end of JSON input
dotnet.js:35Uncaught (in promise) Error: Failed to load config file ./blazor.boot.json SyntaxError: Unexpected end of JSON input SyntaxError: Unexpected end of JSON input

## To Reproduce
Create a new xplat test application

1. Change the AvaloniaVersion to <AvaloniaVersion>11.1.999-cibuild0045495-beta</AvaloniaVersion>
2. Try to run the application

## Expected behavior
run normally

## Screenshots
![image](https://github.com/AvaloniaUI/Avalonia/assets/3688369/a2412ebf-be55-41c7-b223-a142db46714b)

## Environment 

- OS: [ Windows]
- Avalonia-Version: 
11.1.999-cibuild0045495-beta

## MORE 

The reason i wanted to try the latest nightly build was i having another problem with the version 11.1.999-cibuild0043825-beta
If you add Dispatcher.InvokeAsync() it stucks on splash screen. So i wanted to try the 11.1.999-cibuild0045495-beta then i got thiserror. By the way everything works fine with .net 8.0 and stable version 11.0.6. But since my app use some bug fixes for ios i didn't want to use the stable version. If you have any insight about the Dispatcher problem i would like to hear that too. Thank you for the assistance.

**Comment by maxkatz6** [comment]
Add `<WasmRuntimeAssetsLocation>./</WasmRuntimeAssetsLocation>` to your csproj file.
Or even better replace:
```
import { dotnet } from './dotnet.js'
```
with
```
import { dotnet } from './_framework/dotnet.js'
```

In your `main.js` file (only if targeting .NET 8, which is recommended).

**Reply by SilverioMiranda** [reply]
That worked for me

**Comment by maxkatz6** [comment]
This will be documented in a 11.1 changelog. But in general, is caused by .NET 8 WASM SDK breaking changes, which were avoided in 11.0, but not in 11.1.

**Comment by maxkatz6** [comment]
> If you add Dispatcher.InvokeAsync() it stucks on splash screen.

You also should replace `runMainAndExit` with `runMain` in `main.js` file.

**Comment by cangunaydin** [comment]
Hello Max,
Thanks for your quick response, it seems like it is working.
if i change my main.js like this.

``` csharp
import { dotnet } from './_framework/dotnet.js'

const is_browser = typeof window != "undefined";
if (!is_browser) throw new Error(`Expected to be running in a browser`);

const dotnetRuntime = await dotnet
    .withDiagnosticTracing(false)
    .withApplicationArgumentsFromQuery()
    .create();

const config = dotnetRuntime.getConfig();

await dotnetRuntime.runMain(config.mainAssemblyName, [window.location.search]);
```

i get the same problem.

![image](https://github.com/AvaloniaUI/Avalonia/assets/3688369/14594d7a-b8b7-4031-b692-71c3cb7283f1)

if i change the 

```
<link rel="modulepreload" href="./dotnet.js" />
```
to

```
<link rel="modulepreload" href="./_framework/dotnet.js" />
```

then my first problem has been fixed, but dispatcher problem still kind of persists. Dispatcher seems to be working now breakpoint hits it. The problem is, it doesn't change the ui when i raise some changes.
```
public class MainViewModel : ViewModelBase
{
    private string _greeting = "Welcome to Avalonia!";

    public string Greeting
    {
        get => _greeting;
        set => this.RaiseAndSetIfChanged(ref _greeting, value);
    }


    public void LoadPage()
    {
        Dispatcher.UIThread.InvokeAsync(() =>
        {
            Greeting = "Welcome message changed!";
        });
    }
}
```
and this is the mainview.

```
public partial class MainView : UserControl
{
    public MainView()
    {
        InitializeComponent();
        this.AttachedToVisualTree += OnLoaded;
    }

    private void OnLoaded(object? sender, VisualTreeAttachmentEventArgs e)
    {
        var mainViewModel = (MainViewModel)DataContext!;
        mainViewModel.LoadPage();
    }
}
```

maybe it is related with AttachedToVisualTree  event? you can not invoke the dispatcher from AttachedToVisualTree event? cause it seems it works when i add the same method to mainviewmodel constructor?

**Comment by timunie** [comment]
Try Post instead of InvokeAsync

**Comment by cangunaydin** [comment]
Yes it works with Post, @timunie.  thank you. 
Can i ask should i avoid using invokeasync everywhere or is this related only with AttachedToVisualTree event?

**Comment by timunie** [comment]
async will be on a different thread and thus notifications don't work correctly afaik. So better Post unless InvokeAsync reaaaly needed.

**Reply by cangunaydin** [reply]
thanks for the info
can we say this is happening because wasm is single threaded in short?
.net 8 wasm is still single threaded i suppose?
